keadm/cmd/keadm/app/cmd/beta/config.go

```diff
@@ -227,8 +236,62 @@ func GetPauseImage() string {
 	return constants.DefaultPodSandboxImage
 }
 
-// dockerPullImage uses Docker to pull the image
-func dockerPullImage(ctx context.Context, image string, cli *dockerclient.Client) error {
+type ContainerRuntime interface {
+	PullImage(image string) error
+}
+
+func NewContainerRuntime(runtimeType string, endpoint string) (ContainerRuntime, error) {
+	var runtime ContainerRuntime
+	switch runtimeType {
+	case kubetypes.DockerContainerRuntime:
+		// default using docker
+		runtime = &DockerRuntime{}
+	case kubetypes.RemoteContainerRuntime:
+		runtime = &CRIRuntime{
+			endpoint: endpoint,
+		}
+	default:
+		return runtime, fmt.Errorf("unsupport CRI runtime: %s", runtimeType)
+	}
+
+	return runtime, nil
+}
+
+type CRIRuntime struct {
+	endpoint string
+}
+
+func (runtime *CRIRuntime) PullImage(image string) error {
+	imageService, err := remote.NewRemoteImageService(runtime.endpoint, time.Second*10)
+	if err != nil {
+		return err
+	}
+
+	imageSpec := &runtimeapi.ImageSpec{Image: image}
+	status, err := imageService.ImageStatus(imageSpec)
+	if err != nil {
+		return err
+	}
+	if status == nil || status.Id == "" {
+		if _, err := imageService.PullImage(imageSpec, nil, nil); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type DockerRuntime struct {
+}
+
+func (runtime *DockerRuntime) PullImage(image string) error {
+	cli, err := dockerclient.NewClientWithOpts(dockerclient.FromEnv)
+	if err != nil {
+		return fmt.Errorf("init docker client failed: %v", err)
+	}
+
```

Comment:
I think  we should call docker `NegotiateAPIVersion` to solve docker client and server API version mismatch problem. I have encountered this problem ever